From ee0f387d5ddfdfedc7336bf3c90eb3dafd1fa1d8 Mon Sep 17 00:00:00 2001
From: Ben Page <744035+ben-page@users.noreply.github.com>
Date: Thu, 29 Oct 2026 13:47:39 -0500
Subject: [PATCH 1/2] combase/tests: Add tests for
 RoGetParameterizedTypeInstanceIID.

The expected IIDs were computed from hand-written type signatures with an
independent SHA-1 implementation; those for IIterable<String>,
IReference<Point> and TypedEventHandler<UISettings, Object> match the values
in the Windows SDK headers and widl's generated headers.
---
 dlls/combase/tests/roapi.c     | 166 +++++++++++++++++++++++++++++++++
 include/roparameterizediid.idl |   4 +
 2 files changed, 170 insertions(+)

diff --git a/dlls/combase/tests/roapi.c b/dlls/combase/tests/roapi.c
index 70977c9..2324636 100644
--- a/dlls/combase/tests/roapi.c
+++ b/dlls/combase/tests/roapi.c
@@ -27,6 +27,7 @@
 #include "initguid.h"
 #include "roapi.h"
 #include "roerrorapi.h"
+#include "roparameterizediid.h"
 
 #include "wine/test.h"
 
@@ -866,6 +867,170 @@ static void test_RoGetErrorReportingFlags(void)
     ok(flags == RO_ERROR_REPORTING_USESETERRORINFO, "Got unexpected flag %#x.\n", flags);
 }
 
+enum test_type_kind
+{
+    TEST_INTERFACE,
+    TEST_DELEGATE,
+    TEST_PARAMETERIZED_INTERFACE,
+    TEST_PARAMETERIZED_DELEGATE,
+    TEST_RUNTIMECLASS,
+    TEST_RUNTIMECLASS_PARAMETERIZED,
+    TEST_STRUCT,
+    TEST_ENUM,
+    TEST_NOT_FOUND,
+};
+
+struct test_type
+{
+    const WCHAR *name;
+    enum test_type_kind kind;
+    GUID iid;
+    UINT32 count;
+    const WCHAR *names[13];
+};
+
+static const struct test_type test_types[] =
+{
+    {L"Windows.Foundation.Collections.IIterable`1", TEST_PARAMETERIZED_INTERFACE,
+     {0xfaa585ea, 0x6214, 0x4217, {0xaf, 0xda, 0x7f, 0x46, 0xde, 0x58, 0x69, 0xb3}}, 1},
+    {L"Windows.Foundation.Collections.IVector`1", TEST_PARAMETERIZED_INTERFACE,
+     {0x913337e9, 0x11a1, 0x4345, {0xa3, 0xa2, 0x4e, 0x7f, 0x95, 0x6e, 0x22, 0x2d}}, 1},
+    {L"Windows.Foundation.Collections.IMap`2", TEST_PARAMETERIZED_INTERFACE,
+     {0x3c2925fe, 0x8519, 0x45c1, {0xaa, 0x79, 0x19, 0x7b, 0x67, 0x18, 0xc1, 0xc1}}, 2},
+    {L"Windows.Foundation.IReference`1", TEST_PARAMETERIZED_INTERFACE,
+     {0x61c17706, 0x2d65, 0x11e0, {0x9a, 0xe8, 0xd4, 0x85, 0x64, 0x01, 0x54, 0x72}}, 1},
+    {L"Windows.Foundation.TypedEventHandler`2", TEST_PARAMETERIZED_DELEGATE,
+     {0x9de1c534, 0x6ae1, 0x11e0, {0x84, 0xe1, 0x18, 0xa9, 0x05, 0xbc, 0xc5, 0x3f}}, 2},
+    {L"Windows.Foundation.IClosable", TEST_INTERFACE,
+     {0x30d5a829, 0x7fa4, 0x4026, {0x83, 0xbb, 0xd7, 0x5b, 0xae, 0x4e, 0xa9, 0x9e}}},
+    {L"Windows.Foundation.DeferralCompletedHandler", TEST_DELEGATE,
+     {0xed32a372, 0xf3c8, 0x4faa, {0x9c, 0xfb, 0x47, 0x01, 0x48, 0xda, 0x38, 0x88}}},
+    {L"Windows.UI.ViewManagement.UISettings", TEST_RUNTIMECLASS,
+     {0x85361600, 0x1c63, 0x4627, {0xbc, 0xb1, 0x3a, 0x89, 0xe0, 0xbc, 0x9c, 0x55}}, 1,
+     {L"Windows.UI.ViewManagement.IUISettings"}},
+    {L"Windows.Foundation.Collections.StringMap", TEST_RUNTIMECLASS_PARAMETERIZED, {0}, 3,
+     {L"Windows.Foundation.Collections.IMap`2", L"String", L"String"}},
+    {L"Windows.Foundation.Point", TEST_STRUCT, {0}, 2, {L"Single", L"Single"}},
+    {L"Test.Fundamentals", TEST_STRUCT, {0}, 13,
+     {L"UInt8", L"Int16", L"UInt16", L"Int32", L"UInt32", L"Int64", L"UInt64",
+      L"Single", L"Double", L"Boolean", L"Char16", L"String", L"Guid"}},
+    {L"Windows.UI.ViewManagement.UIColorType", TEST_ENUM, {0}, 1, {L"Int32"}},
+    {L"Windows.Storage.FileAttributes", TEST_ENUM, {0}, 1, {L"UInt32"}},
+    {L"Test.NotFound`1", TEST_NOT_FOUND},
+};
+
+static HRESULT STDMETHODCALLTYPE test_locator_Locate(IRoMetaDataLocator *iface, const WCHAR *name,
+                                                     IRoSimpleMetaDataBuilder *builder)
+{
+    const struct test_type *type;
+
+    for (type = test_types; type < test_types + ARRAY_SIZE(test_types); type++)
+        if (!wcscmp(type->name, name)) break;
+    if (type == test_types + ARRAY_SIZE(test_types))
+    {
+        ok(0, "Unexpected type name %s.\n", debugstr_w(name));
+        return E_FAIL;
+    }
+
+    switch (type->kind)
+    {
+    case TEST_INTERFACE:
+        return IRoSimpleMetaDataBuilder_SetWinRtInterface(builder, type->iid);
+    case TEST_DELEGATE:
+        return IRoSimpleMetaDataBuilder_SetDelegate(builder, type->iid);
+    case TEST_PARAMETERIZED_INTERFACE:
+        return IRoSimpleMetaDataBuilder_SetParameterizedInterface(builder, type->iid, type->count);
+    case TEST_PARAMETERIZED_DELEGATE:
+        return IRoSimpleMetaDataBuilder_SetParameterizedDelegate(builder, type->iid, type->count);
+    case TEST_RUNTIMECLASS:
+        return IRoSimpleMetaDataBuilder_SetRuntimeClassSimpleDefault(builder, type->name, type->names[0], &type->iid);
+    case TEST_RUNTIMECLASS_PARAMETERIZED:
+        return IRoSimpleMetaDataBuilder_SetRuntimeClassParameterizedDefault(builder, type->name, type->count,
+                                                                             (const WCHAR **)type->names);
+    case TEST_STRUCT:
+        return IRoSimpleMetaDataBuilder_SetStruct(builder, type->name, type->count, (const WCHAR **)type->names);
+    case TEST_ENUM:
+        return IRoSimpleMetaDataBuilder_SetEnum(builder, type->name, type->names[0]);
+    case TEST_NOT_FOUND:
+        return RO_E_METADATA_NAME_NOT_FOUND;
+    }
+    return E_FAIL;
+}
+
+static const IRoMetaDataLocatorVtbl test_locator_vtbl =
+{
+    test_locator_Locate,
+};
+
+static void test_RoGetParameterizedTypeInstanceIID(void)
+{
+    static const struct
+    {
+        UINT32 count;
+        const WCHAR *names[4];
+        GUID iid;
+    }
+    tests[] =
+    {
+        /* IIterable<String> */
+        {2, {L"Windows.Foundation.Collections.IIterable`1", L"String"},
+         {0xe2fcc7c1, 0x3bfc, 0x5a0b, {0xb2, 0xb0, 0x72, 0xe7, 0x69, 0xd1, 0xcb, 0x7e}}},
+        /* TypedEventHandler<UISettings, Object> */
+        {3, {L"Windows.Foundation.TypedEventHandler`2", L"Windows.UI.ViewManagement.UISettings", L"Object"},
+         {0x2dbdba9d, 0x20da, 0x519d, {0x90, 0x78, 0x09, 0xf8, 0x35, 0xbc, 0x5b, 0xc7}}},
+        /* IMap<IVector<Int32>, String> */
+        {4, {L"Windows.Foundation.Collections.IMap`2", L"Windows.Foundation.Collections.IVector`1", L"Int32", L"String"},
+         {0xd11addfe, 0x3095, 0x5c97, {0x86, 0x4b, 0xa2, 0xd5, 0xff, 0xbb, 0xc3, 0xa0}}},
+        /* IReference<Point> */
+        {2, {L"Windows.Foundation.IReference`1", L"Windows.Foundation.Point"},
+         {0x84f14c22, 0xa00a, 0x5272, {0x8d, 0x3d, 0x82, 0x11, 0x2e, 0x66, 0xdf, 0x00}}},
+        /* IReference<UIColorType> */
+        {2, {L"Windows.Foundation.IReference`1", L"Windows.UI.ViewManagement.UIColorType"},
+         {0xa10cb2c9, 0xa59d, 0x592c, {0xbc, 0x7e, 0x3a, 0x40, 0x33, 0xff, 0xcb, 0x34}}},
+        /* IReference<FileAttributes> */
+        {2, {L"Windows.Foundation.IReference`1", L"Windows.Storage.FileAttributes"},
+         {0x7efefa72, 0xa793, 0x5e0c, {0xb3, 0xa9, 0x0a, 0x43, 0x8b, 0x3e, 0x27, 0xd6}}},
+        /* IVector<IClosable> */
+        {2, {L"Windows.Foundation.Collections.IVector`1", L"Windows.Foundation.IClosable"},
+         {0x1bfca4f6, 0x2c4e, 0x5174, {0x98, 0x69, 0xb3, 0x9d, 0x35, 0x84, 0x8f, 0xcc}}},
+        /* IVector<DeferralCompletedHandler> */
+        {2, {L"Windows.Foundation.Collections.IVector`1", L"Windows.Foundation.DeferralCompletedHandler"},
+         {0x4d3494e5, 0x3f72, 0x52ea, {0xa2, 0x08, 0x2a, 0xcb, 0x0c, 0x50, 0x8a, 0x9e}}},
+        /* IVector<StringMap> */
+        {2, {L"Windows.Foundation.Collections.IVector`1", L"Windows.Foundation.Collections.StringMap"},
+         {0x75b467b3, 0xdce0, 0x5a0a, {0x83, 0x02, 0x82, 0x9f, 0x31, 0xb5, 0xc2, 0x29}}},
+        /* IReference<Test.Fundamentals> */
+        {2, {L"Windows.Foundation.IReference`1", L"Test.Fundamentals"},
+         {0x84012d0e, 0x5d1a, 0x5dec, {0xaf, 0x8e, 0x26, 0x33, 0xbc, 0x49, 0x1e, 0x91}}},
+    };
+    static const WCHAR *missing_arg[] = {L"Windows.Foundation.Collections.IVector`1"};
+    static const WCHAR *extra_arg[] = {L"Windows.Foundation.Collections.IIterable`1", L"String", L"String"};
+    static const WCHAR *not_found[] = {L"Test.NotFound`1", L"String"};
+    IRoMetaDataLocator locator = {&test_locator_vtbl};
+    GUID iid;
+    HRESULT hr;
+    UINT i;
+
+    for (i = 0; i < ARRAY_SIZE(tests); i++)
+    {
+        winetest_push_context("%u", i);
+        memset(&iid, 0xcc, sizeof(iid));
+        hr = RoGetParameterizedTypeInstanceIID(tests[i].count, (const WCHAR **)tests[i].names, &locator, &iid, NULL);
+        todo_wine ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+        todo_wine ok(IsEqualGUID(&iid, &tests[i].iid), "Got unexpected iid %s.\n", wine_dbgstr_guid(&iid));
+        winetest_pop_context();
+    }
+
+    hr = RoGetParameterizedTypeInstanceIID(0, missing_arg, &locator, &iid, NULL);
+    ok(FAILED(hr), "Got unexpected hr %#lx.\n", hr);
+    hr = RoGetParameterizedTypeInstanceIID(ARRAY_SIZE(missing_arg), missing_arg, &locator, &iid, NULL);
+    ok(FAILED(hr), "Got unexpected hr %#lx.\n", hr);
+    hr = RoGetParameterizedTypeInstanceIID(ARRAY_SIZE(extra_arg), extra_arg, &locator, &iid, NULL);
+    ok(FAILED(hr), "Got unexpected hr %#lx.\n", hr);
+    hr = RoGetParameterizedTypeInstanceIID(ARRAY_SIZE(not_found), not_found, &locator, &iid, NULL);
+    ok(FAILED(hr), "Got unexpected hr %#lx.\n", hr);
+}
+
 START_TEST(roapi)
 {
     BOOL ret;
@@ -876,6 +1041,7 @@ START_TEST(roapi)
     test_ActivationFactories();
     test_RoGetAgileReference();
     test_RoGetErrorReportingFlags();
+    test_RoGetParameterizedTypeInstanceIID();
 
     SetLastError(0xdeadbeef);
     ret = DeleteFileW(L"wine.combase.test.dll");
diff --git a/include/roparameterizediid.idl b/include/roparameterizediid.idl
index 85ee57d..49adbe7 100644
--- a/include/roparameterizediid.idl
+++ b/include/roparameterizediid.idl
@@ -93,3 +93,7 @@ interface IRoMetaDataLocator
         IRoSimpleMetaDataBuilder *out
     );
 }
+
+cpp_quote("HRESULT WINAPI RoGetParameterizedTypeInstanceIID(UINT32 name_element_count, const WCHAR **name_elements,")
+cpp_quote("                                                 const IRoMetaDataLocator *meta_data_locator, GUID *iid,")
+cpp_quote("                                                 ROPARAMIIDHANDLE *hiid);")
